diff --git a/src/rpc/virnetsocket.c b/src/rpc/virnetsocket.c
index e8fc2d5f7d..472bd8debf 100644
--- a/src/rpc/virnetsocket.c
+++ b/src/rpc/virnetsocket.c
@@ -1556,6 +1556,13 @@ int virNetSocketGetSELinuxContext(virNetSocket *sock,

     virObjectLock(sock);
     if (getpeercon(sock->fd, &seccon) < 0) {
+        // getpeercon from libselinux uses getsockopt() syscall. Some implementations of getsockopts
+        // return EINVAL errno for unsupported valopt argument instead of ENOPROTOOPT errno.
+        // This fix makes libvirt work with such broken implementations.
+        if (errno == EINVAL) {
+            ret = 0;
+            goto cleanup;
+        }
         if (errno == ENOSYS || errno == ENOPROTOOPT) {
             ret = 0;
             goto cleanup;
