diff --git a/target/i386/cpu.c b/target/i386/cpu.c
index 3f0821c15f..3f9475b485 100644
--- a/target/i386/cpu.c
+++ b/target/i386/cpu.c
@@ -7906,10 +7906,11 @@ void cpu_x86_cpuid(CPUX86State *env, uint32_t index, uint32_t count,
         }
         *edx = env->features[FEAT_1_EDX];
         if (threads_per_pkg > 1) {
             *ebx |= threads_per_pkg << 16;
+            *edx |= CPUID_HT;
         }
         if (cpu->pdcm_on_even_without_pmu) {
             if (!cpu->enable_pmu) {
                 *ecx &= ~CPUID_EXT_PDCM;
             }
@@ -8977,9 +8978,6 @@ void x86_cpu_expand_features(X86CPU *cpu, Error **errp)
         }
     }
 
-    if (x86_threads_per_pkg(&env->topo_info) > 1) {
-        env->features[FEAT_1_EDX] |= CPUID_HT;
-    }
 
     for (i = 0; i < ARRAY_SIZE(feature_dependencies); i++) {
         FeatureDep *d = &feature_dependencies[i];
         if (!(env->features[d->from.index] & d->from.mask)) {
