diff --git a/target/i386/cpu.c b/target/i386/cpu.c
index 78308a82a0..d5f3f5276e 100644
--- a/target/i386/cpu.c
+++ b/target/i386/cpu.c
@@ -2213,6 +2213,7 @@ typedef struct X86CPUDefinition {
     int stepping;
     uint8_t avx10_version;
     FeatureWordArray features;
+    FeatureWordArray filtered_features;
     const char *model_id;
     const CPUCaches *const cache_info;
     /*
@@ -3708,6 +3709,7 @@ static const X86CPUDefinition builtin_x86_defs[] = {
             CPUID_PGE | CPUID_MTRR | CPUID_SEP | CPUID_APIC | CPUID_CX8 |
             CPUID_MCE | CPUID_PAE | CPUID_MSR | CPUID_TSC | CPUID_PSE |
             CPUID_DE | CPUID_FP87,
+        .filtered_features[FEAT_1_EDX] = CPUID_HT,
         .features[FEAT_1_ECX] =
             CPUID_EXT_POPCNT | CPUID_EXT_SSE42 | CPUID_EXT_SSE41 |
             CPUID_EXT_CX16 | CPUID_EXT_SSSE3 | CPUID_EXT_SSE3,
@@ -7735,6 +7737,7 @@ static void x86_cpu_load_model(X86CPU *cpu, const X86CPUModel *model)
                             &error_abort);
     for (w = 0; w < FEATURE_WORDS; w++) {
         env->features[w] = def->features[w];
+        cpu->filtered_features[w] |= def->filtered_features[w];
     }
 
     /* legacy-cache defaults to 'off' if CPU model provides cache info */
