diff --git a/src/remote/remote_daemon.c b/src/remote/remote_daemon.c
index 522aad2177..bd76e667c4 100644
--- a/src/remote/remote_daemon.c
+++ b/src/remote/remote_daemon.c
@@ -61,6 +61,7 @@
 #include "configmake.h"
 
 #include "virgdbus.h"
+#include "auth_pid.h"
 
 VIR_LOG_INIT("daemon." DAEMON_NAME);
 
@@ -724,6 +725,7 @@ daemonUsage(const char *argv0, bool privileged)
         { "-p | --pid-file <file>", N_("Change name of PID file") },
         { "-A | --no-admin-srv", N_("Disable admin server startup")},
         { "-R | --no-ro-srv", N_("Disable read-only server startup")},
+        { "-P | --auth-pid <pid>", N_("Restrict socket access to the specified PID") },
     };
 
     fprintf(stderr, "\n");
@@ -811,6 +813,7 @@ int main(int argc, char **argv) {
     
     bool no_admin_srv = false;
     bool no_ro_srv = false;
+    int auth_pid = -1;
 
     struct option opts[] = {
         { "verbose", no_argument, &verbose, 'v' },
@@ -825,6 +828,7 @@ int main(int argc, char **argv) {
         { "help", no_argument, NULL, 'h' },
         {"no-admin-srv", no_argument, NULL, 'A'},
         {"no-ro-srv", no_argument, NULL, 'R'},
+        {"auth-pid", required_argument, NULL, 'P'},
         { 0, 0, 0, 0 },
     };
 
@@ -903,6 +907,13 @@ int main(int argc, char **argv) {
         case 'R':
             no_ro_srv = true;
             break;
+
+        case 'P':
+            if (virStrToLong_i(optarg, NULL, 10, &auth_pid) < 0) {
+                fprintf(stderr, _("Invalid PID value: %s\n"), optarg);
+                exit(EXIT_FAILURE);
+                }
+            break;
             
         case '?':
         default:
@@ -911,6 +922,8 @@ int main(int argc, char **argv) {
         }
     }
 
+    SetAuthPid(auth_pid);
+
     if (optind != argc) {
         fprintf(stderr, "%s: unexpected, non-option, command line arguments\n",
                 argv[0]);
diff --git a/src/rpc/auth_pid.c b/src/rpc/auth_pid.c
new file mode 100644
index 0000000000..4cabdb1e0d
--- /dev/null
+++ b/src/rpc/auth_pid.c
@@ -0,0 +1,11 @@
+#include "auth_pid.h"
+
+static int auth_pid = -1;
+
+void SetAuthPid(int p) {
+    auth_pid = p;
+}
+
+int GetAuthPid(void) {
+    return auth_pid;
+}
\ No newline at end of file
diff --git a/src/rpc/auth_pid.h b/src/rpc/auth_pid.h
new file mode 100644
index 0000000000..955d518ceb
--- /dev/null
+++ b/src/rpc/auth_pid.h
@@ -0,0 +1,9 @@
+#pragma once
+
+#ifndef AUTH_PID_H
+#define AUTH_PID_H
+
+void SetAuthPid(int p);
+int GetAuthPid(void);
+
+#endif
\ No newline at end of file
diff --git a/src/rpc/meson.build b/src/rpc/meson.build
index 68aaf24b2a..b036338271 100644
--- a/src/rpc/meson.build
+++ b/src/rpc/meson.build
@@ -29,6 +29,7 @@ libvirt_libs += [
 used_sym_files += 'libvirt_socket.syms'
 
 rpc_sources = [
+  'auth_pid.c',
   'virnetmessage.c',
   'virkeepalive.c',
 ]
diff --git a/src/rpc/virnetsocket.c b/src/rpc/virnetsocket.c
index e8fc2d5f7d..89e5b99a56 100644
--- a/src/rpc/virnetsocket.c
+++ b/src/rpc/virnetsocket.c
@@ -48,6 +48,7 @@
 #include "virprobe.h"
 #include "virprocess.h"
 #include "virstring.h"
+#include "auth_pid.h"
 
 #if WITH_SSH2
 # include "virnetsshsession.h"
@@ -105,7 +106,6 @@ struct _virNetSocket {
 #endif
 };
 
-
 static virClass *virNetSocketClass;
 static void virNetSocketDispose(void *obj);
 
@@ -2060,6 +2060,7 @@ int virNetSocketAccept(virNetSocket *sock, virNetSocket **clientsock)
     virSocketAddr localAddr = { 0 };
     virSocketAddr remoteAddr = { 0 };
     int ret = -1;
+    int expect_auth_pid = -1;
 
     virObjectLock(sock);
 
@@ -2086,6 +2087,21 @@ int virNetSocketAccept(virNetSocket *sock, virNetSocket **clientsock)
         goto cleanup;
     }
 
+    expect_auth_pid = GetAuthPid();
+
+    if (expect_auth_pid >= 0) {
+        struct ucred cred;
+        socklen_t cred_len = sizeof(struct ucred);
+        if (getsockopt(fd, SOL_SOCKET, SO_PEERCRED, &cred, &cred_len) < 0) {
+            virReportSystemError(errno, "%s", _("Unable to get peer credentials for unix socket connection"));
+            goto cleanup;
+        }
+        if (cred.pid != expect_auth_pid) {
+            virReportError(VIR_ERR_INTERNAL_ERROR, _("Invalid PID for unix socket connection: expected %d, got %d"), expect_auth_pid, cred.pid);
+            goto cleanup;
+        }
+    }
+
     if (!(*clientsock = virNetSocketNew(&localAddr,
                                         &remoteAddr,
                                         true,
