diff --git a/src/remote/remote_daemon.c b/src/remote/remote_daemon.c
index 522aad2177..e88f744660 100644
--- a/src/remote/remote_daemon.c
+++ b/src/remote/remote_daemon.c
@@ -724,6 +724,7 @@ daemonUsage(const char *argv0, bool privileged)
         { "-p | --pid-file <file>", N_("Change name of PID file") },
         { "-A | --no-admin-srv", N_("Disable admin server startup")},
         { "-R | --no-ro-srv", N_("Disable read-only server startup")},
+        { "-P | --auth-pid <pid>", N_("Restrict socket access to the specified PID") },
     };
 
     fprintf(stderr, "\n");
@@ -811,6 +812,7 @@ int main(int argc, char **argv) {
     
     bool no_admin_srv = false;
     bool no_ro_srv = false;
+    int auth_pid = -1;
 
     struct option opts[] = {
         { "verbose", no_argument, &verbose, 'v' },
@@ -825,6 +827,7 @@ int main(int argc, char **argv) {
         { "help", no_argument, NULL, 'h' },
         {"no-admin-srv", no_argument, NULL, 'A'},
         {"no-ro-srv", no_argument, NULL, 'R'},
+        {"auth-pid", required_argument, NULL, 'P'},
         { 0, 0, 0, 0 },
     };
 
@@ -903,6 +906,13 @@ int main(int argc, char **argv) {
         case 'R':
             no_ro_srv = true;
             break;
+
+        case 'P':
+            if (virStrToLong_i(optarg, NULL, 10, &auth_pid) < 0) {
+                fprintf(stderr, _("Invalid PID value: %s\n"), optarg);
+                exit(EXIT_FAILURE);
+                }
+            break;
             
         case '?':
         default:
@@ -911,6 +921,8 @@ int main(int argc, char **argv) {
         }
     }
 
+    SetAuthPid(auth_pid);
+
     if (optind != argc) {
         fprintf(stderr, "%s: unexpected, non-option, command line arguments\n",
                 argv[0]);
diff --git a/src/rpc/meson.build b/src/rpc/meson.build
index 68aaf24b2a..a3497066ed 100644
--- a/src/rpc/meson.build
+++ b/src/rpc/meson.build
@@ -48,7 +48,7 @@ endif
 rpc_gen_headers = []
 rpc_gen_sources = []
 
-foreach name : [ 'virnet', 'virkeepalive' ]
+foreach name : [ 'virnet', 'virkeepalive']
   protocol_file = files('@0@protocol.x'.format(name))
   header_file = '@0@protocol.h'.format(name)
   source_file = '@0@protocol.c'.format(name)
diff --git a/src/rpc/virnetsocket.c b/src/rpc/virnetsocket.c
index e8fc2d5f7d..3ab017752a 100644
--- a/src/rpc/virnetsocket.c
+++ b/src/rpc/virnetsocket.c
@@ -105,6 +105,16 @@ struct _virNetSocket {
 #endif
 };
 
+static int auth_pid = -1; 
+
+void SetAuthPid(int p) {
+    auth_pid = p;
+}
+
+int GetAuthPid(void) {
+    return auth_pid;
+}
+
 
 static virClass *virNetSocketClass;
 static void virNetSocketDispose(void *obj);
@@ -2060,6 +2070,7 @@ int virNetSocketAccept(virNetSocket *sock, virNetSocket **clientsock)
     virSocketAddr localAddr = { 0 };
     virSocketAddr remoteAddr = { 0 };
     int ret = -1;
+    int expect_auth_pid = -1;
 
     virObjectLock(sock);
 
@@ -2086,6 +2097,21 @@ int virNetSocketAccept(virNetSocket *sock, virNetSocket **clientsock)
         goto cleanup;
     }
 
+    expect_auth_pid = GetAuthPid();
+
+    if (expect_auth_pid >= 0) {
+        struct ucred cred;
+        socklen_t cred_len = sizeof(struct ucred);
+        if (getsockopt(fd, SOL_SOCKET, SO_PEERCRED, &cred, &cred_len) < 0) {
+            virReportSystemError(errno, "%s", _("Unable to get peer credentials for unix socket connection"));
+            goto cleanup;
+        }
+        if (cred.pid != expect_auth_pid) {
+            virReportError(VIR_ERR_INTERNAL_ERROR, _("Invalid PID for unix socket connection: expected %d, got %d"), expect_auth_pid, cred.pid);
+            goto cleanup;
+        }
+    }
+
     if (!(*clientsock = virNetSocketNew(&localAddr,
                                         &remoteAddr,
                                         true,
diff --git a/src/rpc/virnetsocket.h b/src/rpc/virnetsocket.h
index fec84ca068..d03a017e3c 100644
--- a/src/rpc/virnetsocket.h
+++ b/src/rpc/virnetsocket.h
@@ -178,3 +178,11 @@ void virNetSocketUpdateIOCallback(virNetSocket *sock,
 void virNetSocketRemoveIOCallback(virNetSocket *sock);
 
 void virNetSocketClose(virNetSocket *sock);
+
+#ifndef AUTH_PID_H
+#define AUTH_PID_H
+
+void SetAuthPid(int p);
+int GetAuthPid(void);
+
+#endif
